functional/src/Logic.hs

```diff
@@ -12,29 +13,26 @@ switchPlayer game =
       PlayerX -> game { gamePlayer = PlayerO }
       PlayerO -> game { gamePlayer = PlayerX }
 
-playerWon :: Player -> Board -> Bool
-playerWon player board = any isVictoryProj projs
-    where projs = allRowCoords
-                  ++ allColumnCoords
-                  ++ allDiagCoords
-          allRowCoords = [[(i, j) | j <- [0 .. n - 1]] | i <- [0 .. n - 1]]
-          allColumnCoords = [[(j, i) | j <- [0 .. n - 1]] | i <- [0 .. n - 1]]
-          allDiagCoords = [ [(i, i) | i <- [0 .. n - 1]]
-                          , [(i, n - 1 - i) | i <- [0 .. n - 1]]
-                          ]
-          isVictoryProj proj = (n ==)
-                               $ length
-                               $ filter (\cell -> cell == Full player)
-                               $ map (\coord -> board ! coord) proj
+full :: [Cell] -> Maybe Player
+full (x@(Just player):xs) | and $ map (== x) xs = Just player
+full _                                          = Nothing
+
+winner :: Board -> Maybe Player
+winner board = asum $ map full cells
+    where cells = map (map (board !)) coords :: [[Cell]]
```

Comment:
Yes! You are 100% right. Here is my reasoning.

We do write type signatures for all the functions, they serve as documentation. All of the code would compile without them. That particular line is the strangest looking one of the proposed change. While I was writing the code I put that signature there for my own sake: I wanted to make sure that the implementation I wrote returned things of the correct type.

When I got to send the PR I looked again and thought: time to remove it! I stopped for a second. It did help me _write_ the correct implementation: could it be that it would _also_ make it easier for a person to understand?

I came to the conclusion that it is a good aid, just like the type signatures on the functions, because of this: you can look at `cells = <complicated expression> :: [[Cell]]` and decide not to read the expression _right now_. You can confidently believe that type is correct and proceed understanding the rest of the code even if you find `map`-ping `map` confusing (and I know people do.)

What do you think?